### ISSUE TITLE
BOSH deploy failure: observing multiple envoy sidecar processes in permanent D state on diego-cells

### ISSUE THREAD

**Issue by gcapizzi** [cloudfoundry/garden-runc-release#99]
# Description

A redeploy of a diego cellblock failed and when we SSH onto the VM, we observe many garden-init, envoy and some java processes in D state.

```
Task 4368563 | 00:47:58 | Updating instance diego-cell: diego-cell/45a156a0-8a95-4f2a-852e-93762f3aae3f (14) (00:23:30)
                     L Error: Action Failed get_task: Task e975c0e6-fd78-47e3-61a6-10ececb52ea3 result: 1 of 5 post-start scripts failed. Failed Jobs: rep. Successful Jobs: bosh-dns, route_emitter, vxlan-policy-agent, silk-daemon.
Task 4368563 | 00:47:58 | Error: Action Failed get_task: Task e975c0e6-fd78-47e3-61a6-10ececb52ea3 result: 1 of 5 post-start scripts failed. Failed Jobs: rep. Successful Jobs: bosh-dns, route_emitter, vxlan-policy-agent, silk-daemon.
```

It may or may not be important but we also observed that the freezer cgroup for each of the envoy containers had its state set to `FREEZING` (something runc does when deleting containers) - which according to kernel docs should be a transient state but we observe it never changes.

Attempting to kill any garden-init or envoy processes with either SIGTERM or SIGKILL appears to have no effect.

This **may** be related to #96 

# Environment

* garden-runc-release version: (if upgrading, please note both versions; from and to)
* IaaS: AWS
* Stemcell version: xenial 97.17
* Kernel version: 4.15.0-33-generic

# Steps to reproduce

?

# Logs

Logs have been sent to @BooleanCat (they may contain sensitive data).

# Cause

?

Suspicions: we're looking at a line of thought where envoy may be responsible (due to the fact that we have observed envoy d stating before and we have another similar issue where envoy is d state) but have nothing concrete atm.

# Resolution

?

**Comment by teddyking** [comment]
PS maybe relevant - https://www.pivotaltracker.com/n/projects/1158420/stories/159427098 (especially Re: `FREEZING` state). TL;DR totally normal for containers to go into `FREEZING` state as part of a destroy, and thus for them to remain `FREEZING` if a destroy is failing for whatever reason (aka possibly D-state).

**Comment by Callisto13** [comment]
Our top working theory for this problem is documented here: https://www.pivotaltracker.com/story/show/162454024

We have not managed to reproduce, so if anyone encounters this again, PLEASE save us the afflicted VM and do not roll it!

**Comment by Callisto13** [comment]
We have now determined that this is caused by a race in the kernel SRCU (sleepable read-copy update) mechanism.

TLDR:
When doing a rolling deploy on Diego Cells, Garden will be destroy all containers, sidecars and peas. All containers in the affected CFs have associated sidecar containers running Envoy proxy. The Envoy proxy uses the `inotify` file watcher.  When the destroy call comes, all the Envoys will call `inotify_release` en-masse. Part of releasing involves destroying something called 'marks'. This is handled by a kernel worker, which is susceptible to a recently known data race: https://lists.gt.net/linux/kernel/3124443.

We have tested against a kernel patched with a potential fix and have proved that it works. Canonical are now backporting the fix into kernel 4.15. This should happen around March 2019. From then it should make it into our BOSH stemcells.

**Comment by Callisto13** [comment]
Fixed from stemcell v250.29
